### PR TITLE
Classlib: Fix IDict collect, select, reject to keep parent, proto and know

### LINKED ISSUE #2499: collect on IdentityDictionary loses state variables
```
e = (a: 1).parent_((b: 2));
f = e.collect(_+10);
-> ( 'a': 11 )

[e.parent, f.parent]
-> [ ( 'b': 2 ), nil ]
```

`f.parent == nil` seems wrong to me. I would assume "collect" would preserve state variables. 

So now, after https://github.com/supercollider/supercollider/commit/cce6208d9bc2292aac91a8521728a0d0ebbc63f7, the problem is, if a child pattern (such as, in Pchain) calls processRest on an Event, now it uses collect on the Event (which it didn't before), and this throws away the input event's parent.

```
p = Pchain(
	Pfunc { |ev| ev.parent.debug("parent 2"); ev },
	Pfunc { |ev| ev.parent.debug("parent 1"); ev }
).asStream;

e = p.next(().parent_((b: 2)));
parent 1: ( 'b': 2 )
parent 2: nil   // what the...???

e.parent;
-> nil   // what the...???
```

I think the proper fix is to implement "collect" for IdentityDictionary, separately from Dictionary, to migrate the proto and parent into the new dictionary. Possibly this should be done for select, reject etc. I could make a strong case that "collect" should have been doing this for years -- i.e., this bug has been there ever since the proto and parent variables were introduced, and nobody noticed because we don't usually collect on events with parents. 
 
But, this might have a performance impact on anything doing "processRest" on an Event. An alternate would be to define processRest as a no-op for events. It smells funny to me, so I would consider this a last resort, but it might be justifiable. 
 
(So this time, I kept my mouth shut until figuring out that it was my action that exposed the bug (though, if my analysis is correct, the bug is someone else's). But I will note that the trend holds: this bug is really spectacularly obscure -- over an hour to track down -- and I wonder how the hell it is that nobody besides myself has to waste so much time on things like this.)

### PR DESCRIPTION
**Pull request by jamshark70**
Also, clarify documentation and fix a broken link.

Fixes #2499.

### CODE REVIEW

File: HelpSource/Classes/IdentityDictionary.schelp [hunk 1]
```diff
@@ -22,6 +22,52 @@ The contents of a Dictionary are strong::unordered::. You must not depend on the
...
+
+IdentityDictionary has three levels of content: the dictionary itself, a code::proto::, and a code::parent::. code::proto:: and code::parent:: exist primarily for IdentityDictionary's subclass link::Classes/Event::, which uses the code::parent:: to store action functions that will be used when an event is played. Users may put additional default values into the code::proto::.
+
+Looking up a key within a dictionary first checks the dictionary itself. If the key is not found at this level, it looks in the code::proto::, and if still not found, it looks in the parent.
+
+These extra levels are meant for common, default values that should be the same across many dictionary instances.
+
+list::
+## code::proto:: and code::parent:: values are not shown when posting the dictionary.
+## Copying an IdentityDictionary transfers a reference to the parent and proto into the copy. It is a reference, not a copy: changing the copy's parent also affects the parent of the original. This is for efficiency.
```
Comment (telephon) [inline HelpSource/Classes/IdentityDictionary.schelp]: (suggestion:)

this is more efficient than a deep copy, and also allows to inherit the dynamic relation to the parent.

File: HelpSource/Classes/IdentityDictionary.schelp [hunk 2]
```diff
@@ -22,6 +22,52 @@ The contents of a Dictionary are strong::unordered::. You must not depend on the
...
+## Looping operations (link::Classes/Dictionary#-do::, link::Classes/Dictionary#-keysValuesDo::) only touch the main-level values.
```
Comment (telephon) [inline HelpSource/Classes/IdentityDictionary.schelp]: technically this is iteration, no?

File: HelpSource/Classes/IdentityDictionary.schelp [hunk 3]
```diff
@@ -22,6 +22,52 @@ The contents of a Dictionary are strong::unordered::. You must not depend on the
...
+e = IdentityDictionary(8,
+	parent: IdentityDictionary[\b -> 10]
+).put(\a, 5);
+-> IdentityDictionary[ (a -> 5) ]
+
+// But 'b' is still there
+e[\a] + e[\b]
+-> 15
+
+// Looping touches 'a' only
```
Comment (telephon) [inline HelpSource/Classes/IdentityDictionary.schelp]: again, I wouldn't have understood this as looping, but that may only be me.

File: SCClassLibrary/Common/Collections/Dictionary.sc [hunk 4]
```diff
@@ -449,6 +449,23 @@ IdentityDictionary : Dictionary {
...
+	}
+	select { arg function;
+		var res = this.class.new(this.size, proto, parent, know);
+		this.keysValuesDo { arg key, elem; if(function.value(elem, key)) { res.put(key, elem) } }
+		^res;
+	}
+	reject { arg function;
+		var res = this.class.new(this.size, proto, parent, know);
+		this.keysValuesDo { arg key, elem; if(function.value(elem, key).not)
+			{ res.put(key, elem) } }
```
Comment (telephon) [inline SCClassLibrary/Common/Collections/Dictionary.sc]: formatting?

**Comment by snappizz** [comment]
I've pushed to the branch to address @telephon's issues, so this is ready to merge.

**Comment by telephon** [comment]
thanks!

**Pull request merged**